toolchain/driver/tool_runner_base.h

```diff
@@ -0,0 +1,60 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_DRIVER_TOOL_RUNNER_BASE_H_
+#define CARBON_TOOLCHAIN_DRIVER_TOOL_RUNNER_BASE_H_
+
+#include <optional>
+
+#include "common/ostream.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "toolchain/install/install_paths.h"
+
+namespace Carbon {
+
+// Base class that factors out common utilities needed when implementing a
+// runner of an external tool, especially tools part of the LLVM and Clang C++
+// toolchain that Carbon will end up wrapping.
+class ToolRunnerBase {
+ protected:
+  // Construct the tool runner bas.
+  //
+  // If `vlog_stream` is provided, it will be used for `CARBON_VLOG`s. If it is
+  // also equal to `&llvm::errs()`, and so tied to stderr, that will be used by
+  // verbose flag injection helpers in this class.
+  explicit ToolRunnerBase(const InstallPaths* install_paths,
+                          llvm::raw_ostream* vlog_stream = nullptr);
+
+  // Translates `args` into C-string arguments for tool APIs based on `main`.
+  //
+  // Accepts a `tool_name` for logging, and a `tool_path` that will be used as
+  // the first C-string argument to simulate and `argv[0]` entry.
+  //
+  // Accepts a `cstr_arg_storage` that will provide the underlying storage for
+  // the C-strings, and returns a small vector of the C-string pointers. The
+  // returned small vector uses a large small size to allow most common command
+  // lines to avoid extra allocations and growth passes.
+  //
+  // Lastly accepts an optional `verbose_flag`. If provided, and if
+  // `vlog_stream_` is bound to stderr for this instance, the verbose flag will
+  // be injected at the start of the argument list.
+  auto BuildCStrArgs(llvm::StringRef tool_name, llvm::StringRef tool_path,
+                     llvm::ArrayRef<llvm::StringRef> args,
+                     llvm::OwningArrayRef<char>& cstr_arg_storage,
+                     std::optional<llvm::StringRef> verbose_flag = {})
```

Comment:
It feels like with `cstr_arg_storage` in the middle is a little unusual. Style recommends putting output parameters after inputs, which of course gets awkward with the default parameter. A few different thoughts:

- It looks like verbose_flag might always be specified as "-v", could that be made required and moved before `cstr_arg_storage`?
- Alternately, maybe this could return `std::pair<llvm::OwningArrayRef<char>, llvm::SmallVector<const char*, 64>>`, or perhaps a `struct`?

Reply:
Made the verbose flag not have a default argument and moved it before `args`.